system/Files/FileCollection.php

```diff
@@ -0,0 +1,396 @@
+		}
+
+		return array_filter($files, function ($value) use ($pattern) {
+			return (bool) preg_match($pattern, basename($value));
+		});
+	}
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Loads the Filesystem helper and adds any initial files.
+	 *
+	 * @param string[] $files
+	 */
+	public function __construct(array $files = [])
+	{
+		helper(['filesystem']);
+
+		$this->add($files)->define();
+	}
+
+	/**
+	 * Applies any initial inputs after the constructor.
+	 * This method is a stub to be implemented by child classes.
+	 *
+	 * @return void
+	 */
+	protected function define(): void
+	{
+	}
+
+	/**
+	 * Optimizes and returns the current file list.
+	 *
+	 * @return string[]
+	 */
+	public function get(): array
+	{
+		$this->files = array_unique($this->files, SORT_STRING);
+		sort($this->files, SORT_STRING);
+
+		return $this->files;
+	}
+
+	/**
+	 * Sets the file list directly, files are still subject to verification.
+	 * This works as a "reset" method with [].
+	 *
+	 * @param string[] $files The new file list to use
+	 *
+	 * @return $this
+	 */
+	public function set(array $files)
+	{
+		$this->files = [];
+
+		return $this->addFiles($files);
+	}
+
+	/**
+	 * Adds an array/single file or directory to the list.
+	 *
+	 * @param string|string[] $paths
+	 * @param boolean         $recursive
+	 *
+	 * @return $this
+	 */
+	public function add($paths, bool $recursive = true)
+	{
+		if (! is_array($paths))
+		{
+			$paths = [$paths];
+		}
+
+		foreach ($paths as $path)
+		{
+			if (! is_string($path))
+			{
+				throw new InvalidArgumentException('FileCollection paths must be strings.');
+			}
+
+			// Test for a directory
+			try
+			{
+				$directory = self::resolveDirectory($path);
+			}
+			catch (FileException $e)
+			{
+				return $this->addFile($path);
+			}
+
+			$this->addDirectory($path, $recursive);
+		}
+
+		return $this;
+	}
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Verifies and adds files to the list.
+	 *
+	 * @param string[] $files
+	 *
+	 * @return $this
+	 */
+	public function addFiles(array $files)
+	{
+		foreach ($files as $file)
+		{
+			$this->addFile($file);
+		}
+
+		return $this;
+	}
+
+	/**
+	 * Verifies and adds a single file to the file list.
+	 *
+	 * @param string $file
+	 *
+	 * @return $this
+	 */
+	public function addFile(string $file)
+	{
+		$this->files[] = self::resolveFile($file);
+
+		return $this;
+	}
+
+	/**
+	 * Removes files from the list.
+	 *
+	 * @param string[] $files
+	 *
+	 * @return $this
+	 */
+	public function removeFiles(array $files)
+	{
+		$this->files = array_diff($this->files, $files);
+
+		return $this;
+	}
+
+	/**
+	 * Removes a single file from the list.
+	 *
+	 * @param string $file
+	 *
+	 * @return $this
+	 */
+	public function removeFile(string $file)
+	{
+		return $this->removeFiles([$file]);
+	}
+
+	//--------------------------------------------------------------------
```

Comment:
I love the dividers! But if you want them gone I can strip them. I like the way they allow for grouping methods/properties across large classes visually.

Reply:
I still like "headers" with those where the sections are titled, just not a fan of them separating every method anymore. 

```
//--------------------------------------------------------------------
// Scope Functions
//--------------------------------------------------------------------
```

And maybe that's overkill but sounds like it serves the same purpose to what you were doing here. I am open to discussion on this, of course. Especially for the use case you were doing. I didn't realize it was a section divider since it looked the same as the ones I originally carried over from CI3 code, that most IDE's now have the option to insert for you I think.

Reply:
I prefer dividers for "header" like purposes, but I use `=` for "bolder" appeal.
```
//====================================
// The header name
//====================================
```

Reply:
So in this case do you believe pushing forward to headers is in order, or pull the dividers altogether?

Reply:
I removed some dividers and switched the remaining to headers. I stuck with the single dash for now since that is what we have elsewhere in **system** but I wouldn't be opposed to a sweeping redo around dividers in general.